File: kata-8-kyu/the-feast-of-many-beasts/python/the_feast_of_many_beasts_test.py
# + =====================================================================================================================
# +
# + 8 kyu - The Feast of Many Beasts  [ ID: 5aa736a455f906981800360d ] (the-feast-of-many-beasts)
# + URL: https://www.codewars.com/kata/5aa736a455f906981800360d
# + Category: REFERENCE  |  Tags: STRINGS | FUNDAMENTALS
# +
# + =====================================================================================================================

import math
import random
import codewars_test as test
from the_feast_of_many_beasts import feast


def author_solution(beast, dish):
    return beast[0] == dish[0] and beast[-1] == dish[-1]


@test.describe("Fixed Tests")
def fixed_tests():
    @test.it("Basic Test Cases")
    def basic_test_cases():
        test.assert_equals(feast("great blue heron", "garlic naan"), True)
        test.assert_equals(feast("chickadee", "chocolate cake"), True)
        test.assert_equals(feast("brown bear", "bear claw"), False)
        test.assert_equals(feast("marmot", "mulberry tart"), True)
        test.assert_equals(feast("porcupine", "pie"), True)
        test.assert_equals(feast("cat", "yogurt"), False)
        test.assert_equals(feast("electric eel", "lasagna"), False)
        test.assert_equals(feast("slow loris", "salsas"), True)
        test.assert_equals(feast("ox", "orange lox"), True)
        test.assert_equals(feast("blue-footed booby", "blueberry"), True)


@test.describe("Random Tests")
def random_tests():
    def make_string(minimum, maximum):
        array = []
        possible = "abcdefghijklmnopqrstuvwxyz"
        length = math.ceil((random.random() * maximum) + minimum)
        for _ in range(0, length):
            array.append(possible[math.floor(random.random() * len(possible))])
        return "".join(array)

    for _ in range(100):
        beast = make_string(2, 40)
        dish = make_string(2, 40)
        if random.random() < 0.6:
            dish = dish[:-1] + beast[1]

        @test.it(f"testing for feast({beast}, {dish})")
        def test_case():
            test.assert_equals(feast(beast, dish), author_solution(beast, dish))

File: kata-8-kyu/the-feast-of-many-beasts/typescript/theFeastOfManyBeasts.Test.ts
//+ ====================================================================================================================
//+
//+ 8 kyu - The Feast of Many Beasts  [ ID: 5aa736a455f906981800360d ] (the-feast-of-many-beasts)
//+ URL: https://www.codewars.com/kata/5aa736a455f906981800360d
//+ Category: REFERENCE  |  Tags: STRINGS | FUNDAMENTALS
//+
//+ ====================================================================================================================

import { assert } from "vitest"
import { feast } from "./theFeastOfManyBeasts"

describe("Testing...", function () {
  it("Basic fixed tests", function () {
    assert.isTrue(feast("great blue heron", "garlic naan"), "Should be true")
    assert.isTrue(feast("chickadee", "chocolate cake"), "Should be true")
    assert.isFalse(feast("brown bear", "bear claw"), "Should be false")
    assert.isTrue(feast("marmot", "mulberry tart"), "Should be true")
    assert.isTrue(feast("porcupine", "pie"), "Should be true")
    assert.isFalse(feast("cat", "yogurt"), "Should be false")
    assert.isFalse(feast("electric eel", "lasagna"), "Should be false")
    assert.isTrue(feast("slow loris", "salsas"), "Should be true")
    assert.isTrue(feast("ox", "orange lox"), "Should be true")
    assert.isTrue(feast("blue-footed booby", "blueberry"), "Should be true")
  })
})

describe("Random Tests", () => {
  for (let i = 0; i < 100; i++) {
    const a = makeString()
    let b = makeString()
    if (Math.random() < 0.5) {
      b = a[0] + b + a.slice(-1)
    }
    it(`Testing for beast = ${JSON.stringify(a)}, dish = ${JSON.stringify(b)}`, function () {
      assert.strictEqual(feast(a, b), authorSolution(a, b))
    })
  }
})

const authorSolution = (beast: string, dish: string): boolean => {
  return beast[0] === dish[0] && beast[beast.length - 1] === dish[dish.length - 1]
}

function makeString(): string {
  const len: number = Math.ceil(Math.random() * 40 + 2)
  return [...Array(len)].map(() => String.fromCharCode(Math.floor(Math.random() * 26) + 97)).join("")
}
